src/Raven.Server/Documents/ShardedHandlers/ShardedQueriesHandler.cs

```diff
@@ -0,0 +1,843 @@
+                                           e.ToString();
+                        }
+                        tracker.Query = errorMessage;
+                        if (TrafficWatchManager.HasRegisteredClients)
+                            AddStringToHttpContext(errorMessage, TrafficWatchChangeType.Queries);
+                    }
+                    throw;
+                }
+            }
+        }
+        
+        /// <summary>
+        /// TrafficWatchQuery writes query data to httpContext
+        /// </summary>
+        /// <param name="indexQuery"></param>
+        private void TrafficWatchQuery(IndexQueryServerSide indexQuery)
+        {
+            var sb = new StringBuilder();
+            // append stringBuilder with the query
+            sb.Append(indexQuery.Query);
+            // if query got parameters append with parameters
+            if (indexQuery.QueryParameters != null && indexQuery.QueryParameters.Count > 0)
+                sb.AppendLine().Append(indexQuery.QueryParameters);
+            AddStringToHttpContext(sb.ToString(), TrafficWatchChangeType.Queries);
+        }
+
+        /// <summary>
+        /// A struct that we use to hold state and break the process
+        /// of handling a sharded query into distinct steps
+        /// </summary>
+        public class ShardedQueryProcessor : IComparer<BlittableJsonReaderObject>, IDisposable
+        {
+            private readonly TransactionOperationContext _context;
+            private readonly ShardedQueriesHandler _parent;
+            private readonly IndexQueryServerSide _query;
+            private readonly Dictionary<int, ShardedQueryCommand> _commands;
+            private readonly HashSet<int> _filteredShardIndexes;
+            private readonly ShardedQueryResult _result;
+            private static readonly DynamicJsonValue DummyDynamicJsonValue = new();
+            public const string OrderByMetadataField = "@order-by-fields";
+
+            // User should also be able to define a query parameter ("Sharding.Context") which is an array
+            // that contains the ids whose shards the query should be limited to. Advanced: Optimization if user
+            // want to run a query and knows what shards it is on. Such as:
+            // from Orders where State = $state and User = $user where all the orders are on the same share as the user
+
+            // - TODO: Sharding - etag handling!
+            public ShardedQueryProcessor(TransactionOperationContext context,ShardedQueriesHandler parent,IndexQueryServerSide query)
+            {
+                _context = context;
+                _result = new ShardedQueryResult();
+                _parent = parent;
+                _query = query;
+                _commands = new Dictionary<int, ShardedQueryCommand>();
+                if (_query.QueryParameters != null && _query.QueryParameters.TryGet("Sharding.Context", out BlittableJsonReaderArray filter))
```

Comment:
if we want to support this let's make `Sharding.Context` a const

Reply:
Ping

Reply:
better to use `@` in a name like we do have for `@metadata`

Reply:
added an issue and a TODO